### ISSUE TITLE
Memory Leaks after MenuStrip.Show()

### ISSUE THREAD

**Issue by zzbbcc** [davidegironi/advanceddatagridview#41]
The MenuStrip object will not be released by GC after executing MenuStrip method "public void Show(Control control, int x, int y, IEnumerable<DataGridViewCell> vals)".

**Comment by davidegironi** [comment]
The MenuStrip object it's created once in the ColumnHeaderCell costructor, do I miss something?
Please provie me a Memory Usage snapshot, an example, or something that helps me solve this issue. Thanks!

**Comment by zzbbcc** [comment]
I add a project TestForm to the fork.
https://github.com/zzbbcc/advanceddatagridview

A big array added to MenuStrip to larger the memory allocation.

Reproduce steps:
Scenario 1
1. Click button "FormDGV.Show()".
2. Click button "Set Table".
3. Close FormDGV.
4. Click button GC.Collect, MenuStrip collected by GC.

Scenario 2
1. Click button "FormDGV.Show()".
2. Click button "Set Table".
3. Click MenuStrip to fire its Show() method.
4. Close FormDGV.
5. Click button GC.Collect, MenuStrip isn't collected by GC.
Memory usage of MenuStrip increased on Scenario 2 occurred repeatedly.

**Comment by davidegironi** [comment]
Thank you zzbbcc.

I think this leak may be related to the Dispose method implementation of DataTable and DataRow.
Indeed if you take a Memory Usage snapshot using the Diagnostic Tool, you will find DataRow objects growing in number.
I've slightly edited you Test form, you can find it here: https://filebin.net/oc3lf2wm30ded5tu.
Excuse me but I'm out of my main PC with my github credential, so I will not push this here - also I may not reply for the next week. You will find test for FormDGV (with AdvancedDataGridView) and FormGV (with standard .NET DataGridView).

Garbage collector is being called each 5sec (timer1.Interval of FormMain).

Also in FormMain
//set this to false to give each windows the time to show and close, note that true needs a bit of RAM if many Forms are created
private static bool SpeedTest = true;
//number of forms created
private static int NumWin = 10;
//set this to true to use the FormMain DataTable, false will create one DataTable for each Form.
private static bool LoadDt = true;

Leave LoadDt to false, and look at the Memory Usage Snapshot number of DataRow growing up.

Of course you will notice that more memory is being used by AdvancedDataGridView compared to the standard one, but it's then released.

Please let me know your opinion.

**Comment by zzbbcc** [comment]
Thank you for checking.
It's interesting to know DataRow do leaks. But I also notice MenuStrip may cause leaks larger than DataRow.
I add button "FormDGV Show() + MenuStrip.Show()" to FormMain. And I add a method "ShowAllMenuStrip" to AdvancedDataGridView to bring up all MenuStrip after table loaded. Through diagnostics tool I found TreeNode and Zuby.ADGV.TreeNodeItemSelector growth fast and not released after FormDGV closed.

**Comment by davidegironi** [comment]
So, we have now removed the DataRow leak. Now we are getting in the AdvancedDataGridView leak. Today I've little time to test it.
I've noticed ColumnHeaderCell events never get disconnected when on object is closed, this may cause the GD not to collect the object.
Also pragmatically call a Dispose on MenuStrip may help.

Try adding this on the AdvancedDataGridView.cs file and let me know.
Thank you, this will help other people projects too.

```
        protected override void OnHandleDestroyed(EventArgs e)
        {
            foreach (DataGridViewColumn column in Columns)
            {
                ColumnHeaderCell cell = column.HeaderCell as ColumnHeaderCell;
                if (cell != null)
                {
                    cell.SortChanged -= Cell_SortChanged;
                    cell.FilterChanged -= Cell_FilterChanged;
                    cell.FilterPopup -= Cell_FilterPopup;
                    MenuStrip menustrip = cell.MenuStrip;
                    menustrip.Dispose();
                }
            }

            base.OnHandleDestroyed(e);
        }
```

**Comment by zzbbcc** [comment]
I did tried to dispose MenuStrip on ColumnRemoved before, and it fixed the leaks. But it somehow throw exception at MenuStrip.designer.Dispose() on main process closing with error message "Invalid Parameter". The weird thing is the exception not always happened. Currently I suppress the exception so user won't see the error.

**Comment by davidegironi** [comment]
I made very few tests, more needs to be done, but OnHandleDestroyed i don't get that exception, although if this works on OnColumnRemoved it's maybe better.
Also the events detach is recomended, so I will add this somewhere.
I will try to fix it too. I think I'll have time within a few weeks. If you find a fix without the try catch trick, let me know here.

**Comment by davidegironi** [comment]
Hello @zzbbcc ,
finally I've got time to pack code and test the main code, please test this source: https://filebin.net/s94648vvqu4yooig
It adds the Memory test button.

**Comment by zzbbcc** [comment]
Hi @davidegironi ,
that's great to hear it. I will put this to my project and observe whether the exception still.
Thank you.

**Comment by davidegironi** [comment]
Thank you! I'll wait for your feedback to publish the new version.

**Comment by davidegironi** [comment]
Hi @zzbbcc , have you check it?

**Comment by zzbbcc** [comment]
Hi @davidegironi ,
Yes, I don't receive any exception from then on.

**Comment by davidegironi** [comment]
Thank you!
Fixed on version 1.1.x.8, commited to master, published on nuget.
